ast/ast.go

```diff
@@ -97,16 +97,38 @@ func (Insert) isStatement()             {}
 func (Delete) isStatement()             {}
 func (Update) isStatement()             {}
 
-// QueryExpr represents set operator operands.
+// QueryExpr represents query expression, which can be body of QueryStatement or ParenTableExpr.
+// Select and FromQuery are leaf QueryExpr and others wrap other QueryExpr.
 type QueryExpr interface {
 	Node
 	isQueryExpr()
 }
 
 func (Select) isQueryExpr()        {}
+func (Query) isQueryExpr()         {}
+func (FromQuery) isQueryExpr()     {}
 func (SubQuery) isQueryExpr()      {}
 func (CompoundQuery) isQueryExpr() {}
 
+// AllOrDistinct represents ALL or DISTINCT node.
```

Comment:
IMO, this should be `type AllOrDistinct string` instead, and place it to `const.go`.

Reply:
Also, this can be replace a `Distinct` flag in `CompoundQuery`.

Reply:
My thought is that:

* typed string  isn't well compatible with current doc comment and `sql.go` and `ast.go`.
* typed string can't prevent wrong value.

Reply:
Note: I want to reuse it in pipe operators and GQL statements.

Reply:
It is unclear to me. Please explain it.

> - typed string isn't well compatible with current doc comment, `sql.go`, and `ast.go`.
>   * It spreads logic to many nodes.

No AST node definition is needed to represent one finite type of keyword.
In addition, from my research of the GQL syntax, I am sure that constants of typed strings are sufficient.

Finally, I will make my opinion clear: We MUST use typed string here.

Reply:
I agree it is not required to introduce new AST node type, but it can be a design choice so I haven't yet agree your comment.
We should talk about pros/cons.

> - typed string isn't well compatible with current doc comment, sql.go, and ast.go.
>   - It spreads logic to many nodes.
> It is unclear to me. Please explain it.

1. `SQL()`
Currently, Go `text/format` notation is not needed to be formal, but we need to write `SQL()` equivalent of 
```
{{if .AllOrDistinct}}{{string(.AllOrDistinct)}} {{end}}
```
to many places. It seems `sqlOpt` is easy to write.

2. `Pos()`, `End()`

Actually, `AllOrDistinct` is not necessary to define `pos` and `end` because it does not appear on node boundaries, but let's think about which one we should use if we were to unify either node interface or typed string.
The general pattern for using typed string is that a separate `token.Pos` is required to implement `pos` and `end`, and their code style are not standarized. (Need to store `pos` or `end`?)

https://github.com/cloudspannerecosystem/memefish/blob/bdde76132d427d61b5d32c07cc67e884c66742f5/ast/ast.go#L2146-L2155

https://github.com/cloudspannerecosystem/memefish/blob/bdde76132d427d61b5d32c07cc67e884c66742f5/ast/ast.go#L754-L763

I am doubting that typed string is a good pattern because a single value is no longer self-contained.

Reply:
(It is because `OnDeleteAction` is a constructed by multi token, so we don't use the first token pos, so it seems that we can use `DirEnd` if needed.)